core/wait/wait.go

```diff
@@ -12,72 +12,77 @@ import (
 
 var RetryHttpErrorStatusCodes = []int{http.StatusBadGateway, http.StatusGatewayTimeout}
 
-type WaitFn func() (res interface{}, done bool, err error)
-
-type Handler struct {
-	fn                WaitFn
+// AsyncActionCheck reports whether a specific async action has finished.
+//   - waitFinished == true if the async action is finished, false otherwise.
+//   - response contains data regarding the current state of the resource targeted by the async action (if applicable). resource != nil if waitFinished == true.
+//   - err != nil if there was an error checking if the async action finished, or if it finished unsuccessfully.
+type AsyncActionCheck[T any] func() (waitFinished bool, response *T, err error)
+
+// AsyncActionHandler handles waiting for a specific async action to be finished.
+type AsyncActionHandler[T any] struct {
+	checkFn           AsyncActionCheck[T]
 	sleepBeforeWait   time.Duration
 	throttle          time.Duration
 	timeout           time.Duration
-	tempErrRetryLimit int
+	retryLimitTempErr int
 }
 
-// New creates a new Wait instance
-func New(f WaitFn) *Handler {
-	return &Handler{
-		fn:                f,
+// New initializes an AsyncActionHandler
+func New[T any](f AsyncActionCheck[T]) *AsyncActionHandler[T] {
+	return &AsyncActionHandler[T]{
+		checkFn:           f,
 		sleepBeforeWait:   0 * time.Second,
 		throttle:          5 * time.Second,
 		timeout:           30 * time.Minute,
-		tempErrRetryLimit: 5,
+		retryLimitTempErr: 5,
 	}
 }
 
-// SetThrottle sets the duration between func triggering
-func (w *Handler) SetThrottle(d time.Duration) error {
-	if d == 0 {
-		return fmt.Errorf("throttle can't be 0")
-	}
-	w.throttle = d
-	return nil
+// SetThrottle sets the duration between func triggering.
```

Comment:
Maybe we could make this more explicit? Something like

```suggestion
// SetThrottle sets the time interval between each checkFn triggering.
```

Reply:
Reworded it a bit